ob1k-cache/src/main/java/com/outbrain/ob1k/cache/RefreshLoadingCacheDelegate.java

```diff
@@ -0,0 +1,268 @@
+package com.outbrain.ob1k.cache;
+
+import com.google.common.collect.Iterables;
+import com.outbrain.ob1k.concurrent.ComposableFuture;
+import com.outbrain.swinfra.metrics.api.Counter;
+import com.outbrain.swinfra.metrics.api.MetricFactory;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+/**
+ * A wrapper of LoadingCacheDelegate that adds the ability to refresh cache values in the configured interval on access
+ */
+public class RefreshLoadingCacheDelegate<K, V> implements TypedCache<K, V> {
+
+  private final LoadingCacheDelegate<K, ValueWithWriteTime<V>> cache;
+  private final InternalCacheLoader internalCacheLoader;
+  private final String cacheName;
+  private final long loadDuration;
+  private final TimeUnit loadTimeUnit;
+  private final long refreshAfterWriteDuration;
+
+  private final ConcurrentMap<K, Boolean> refreshingKeys;
+
+  private final Counter refreshes;
+  private final Counter refreshErrors;
+  private final Counter refreshTimeouts;
+
+  public RefreshLoadingCacheDelegate(final TypedCache<K, ValueWithWriteTime<V>> cache, final CacheLoader<K, V> loader, final String cacheName, final MetricFactory metricFactory,
+                                     final long duration, final TimeUnit timeUnit, final boolean failOnError,
+                                     final long refreshAfterWriteDuration, final TimeUnit refreshAfterWriteUnit) {
+    this.cacheName = cacheName;
+    this.internalCacheLoader = new InternalCacheLoader(loader);
+    this.cache = new LoadingCacheDelegate<>(cache, internalCacheLoader, cacheName, metricFactory, duration, timeUnit, failOnError);
+    this.loadDuration = duration;
+    this.loadTimeUnit = timeUnit;
+    this.refreshAfterWriteDuration = refreshAfterWriteUnit.toMillis(refreshAfterWriteDuration);
+
+    refreshingKeys = new ConcurrentHashMap<>();
+
+    if (metricFactory != null) {
+      metricFactory.registerGauge("RefreshLoadingCacheDelegate." + cacheName, "refreshMapSize", refreshingKeys::size);
+
+      refreshes = metricFactory.createCounter("RefreshLoadingCacheDelegate." + cacheName, "refreshes");
+      refreshErrors = metricFactory.createCounter("RefreshLoadingCacheDelegate." + cacheName, "refreshErrors");
+      refreshTimeouts = metricFactory.createCounter("RefreshLoadingCacheDelegate." + cacheName, "refreshTimeouts");
+    } else {
+      refreshes = null;
+      refreshErrors = null;
+      refreshTimeouts = null;
+    }
+  }
+
+  @Override
+  public ComposableFuture<V> getAsync(final K key) {
+    ComposableFuture<ValueWithWriteTime<V>> futureValue = cache.getAsync(key);
+    futureValue.consume(value -> {
+      if (value.isSuccess() && shouldRefresh(value.getValue(), System.currentTimeMillis())) {
```

Comment:
use a time supplier instead of System.currentTimeMillis() and avoid Thread.sleep in tests.